server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/handlers/HeartRateServiceRequestHandler.java

```diff
@@ -0,0 +1,143 @@
+     */
+    @Override
+    public void onServiceRemoved() {
+        heartRateSensorListener.unregisterSubscriber(this);
+        registeredDeviceStorage.removeAllDevices();
+    }
+
+    /**
+     * Unregister device from notifications about heart rate characteristic changed.
+     * @param device connected device
+     */
+    @Override
+    public void onDeviceDisconnected(BluetoothDevice device) {
+        registeredDeviceStorage.removeDevice(device);
+    }
+
+    /**
+     * Determine wether remote device is registered for notifications about Heart Rate Measurement characteristic change or not.
+     *
+     * <p>Read descriptor request for Heart Rate service is the way for remote device to check status of registration
+     * to Heart Rate Measurement characteristic.
+     *
+     * @param device the remote device that has requested the read operation
+     * @param descriptor descriptor to be read
+     * @param offset offset into the value of the descriptor
+     * @return {@link BluetoothGattDescriptor#ENABLE_NOTIFICATION_VALUE} if device is registered for notifications,
+     * {@link BluetoothGattDescriptor#DISABLE_NOTIFICATION_VALUE} otherwise
+     */
+    @Override
+    public byte[] onDescriptorRead(BluetoothDevice device, BluetoothGattDescriptor descriptor, int offset) {
+        if (registeredDeviceStorage.contains(device)) {
+            return BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE;
+        } else {
+            return BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE;
+        }
+    }
+
+    /**
+     * Register (if given value is {@link BluetoothGattDescriptor#ENABLE_NOTIFICATION_VALUE}) or
+     * unregister (if given value is {@link BluetoothGattDescriptor#DISABLE_NOTIFICATION_VALUE})
+     * remote device for notifications about Heart Rate Measurement characteristic change.
+     *
+     * <p>Write descriptor request for Heart Rate service is the way for remote device to subscribe/unsubscribe
+     * for Heart Rate Measurement characteristic change notification.
+     *
+     * @param device the remote device that has requested the read operation
+     * @param descriptor descriptor to be read
+     * @param offset offset into the value of the descriptor
+     */
+    @Override
+    public void onDescriptorWrite(BluetoothDevice device, BluetoothGattDescriptor descriptor, int offset, byte[] value) {
+        if (Arrays.equals(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE, value)) {
+            Log.d(TAG, String.format("Subscribe device %s to notifications", device));
+            registeredDeviceStorage.addDevice(device);
+        } else if (Arrays.equals(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE, value)) {
+            Log.d(TAG, String.format("Unsubscribe device %s from notifications", device));
+            registeredDeviceStorage.removeDevice(device);
+        }
+    }
+
+    /**
+     * Callback to listen to {@link HeartRateSensorListener} about heart rate value change.
```

Comment:
I'd rephrase this line a bit. Maybe: "Called when the heart rate value was changed"?